TypeCobol.Analysis/Dfa/DataFlowGraphBuilder.cs

```diff
@@ -0,0 +1,486 @@
+                    }
+                }
+                IsGenSetCalculated = true;
+            }
+        }
+
+        /// <summary>
+        /// Build the KILL set for each basic block.
+        /// The Kill set represents those variables defined outside the basic block which also have definitions inside the block.
+        /// Those definitions outside the block are destroyed by the redefinitions inside the block.
+        /// </summary>
+        public void ComputeKillSet()
+        {
+            lock (this)
+            {
+                if (IsKillSetCalculated)
+                    return;
+                //prerequisites: Compute the GEN set
+                ComputeGenSet();
+                foreach (var block in Cfg.AllBlocks)
+                {
+                    if (block.Instructions == null || block.Data == null || block.Data.DefCount == 0)
+                        continue;//No Def variables in the block.
+
+                    //Allocate a temporary KILL set.
+                    BitSet kill = new BitSet(DefList.Count);
+
+                    //for each definition outside the basic block
+                    foreach (var def in DefList)
+                    {
+                        if (def.BlockIndex == block.Index)
+                            continue;//Inside the current block
+                        int defIndex;
+                        if (block.Data.GenVariableDictionary.TryGetValue(def.Variable, out defIndex))
+                        {//The def variable is in the Block's GEN, kill the definition outside the block.
+                            kill.Set(def.Index);
+                        }
+                    }
+                    block.Data.Kill = kill;
+                }
+                IsKillSetCalculated = true;
+            }
+        }
+
+        /// <summary>
+        /// Determine if InOuts sets as been calculated
+        /// </summary>
+        public bool IsInOutSetCalculated
+        {
+            get;
+            internal set;
+        }
+
+        /// <summary>
+        /// Initialize the Computation of IN and OUT Sets
+        /// </summary>
+        private void InitializeInOutSetsComputation()
+        {
+            //prerequisites: Compute the KILL sets which will force GEN sets calculation.
+            ComputeKillSet();
+
+            //Initialization
+            foreach (var block in Cfg.AllBlocks)
+            {
+                if (block.Data == null)
+                {
+                    block.Data = new DfaBasicBlockInfo<V>(DefList.Count);
+                }
+                else
+                {
+                    if (block.Data.In == null)
+                        block.Data.In = new BitSet(DefList.Count);
+                    else
+                        block.Data.In.Clear();
+                    if (block.Data.Out == null)
+                        block.Data.Out = new BitSet(DefList.Count);
+                    else
+                        block.Data.Out.Clear();
```

Comment:
This initalization is only for In Out sets calculation, the constructor of DfaBasicBlockInfo allocates In and Out storages.

Reply:
There are two constructors for DfaBasicBlockInfo, the default one does not allocate sets because it doesn't know the cardinality. I've removed the tests `if (block.Data.In == null)` and `if (block.Data.Out == null)` and added an assertion to check that both sets are null.